core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -533,16 +534,18 @@ private void removeStateFromCluster(Collection<? extends IracManagerKeyInfo> sta
       IntSet segments = mutableEmptySet();
       LocalizedCacheTopology cacheTopology = clusteringDependentLogic.getCacheTopology();
       Set<Address> owners = new HashSet<>(cacheTopology.getMembers().size());
-      for (IracManagerKeyInfo state : stateToCleanup) {
+      List<IracManagerKeyInfo> keysToCleanup = new ArrayList<>(stateToCleanup.size());
+      for (IracManagerKeyState state : stateToCleanup) {
+         keysToCleanup.add(state.getKeyInfo());
          if (segments.add(state.getSegment())) {
             owners.addAll(cacheTopology.getSegmentDistribution(state.getSegment()).writeOwners());
          }
       }
 
       if (!segments.isEmpty()) {
-         IracCleanupKeysCommand cmd = commandsFactory.buildIracCleanupKeyCommand(stateToCleanup);
+         IracCleanupKeysCommand cmd = commandsFactory.buildIracCleanupKeyCommand(keysToCleanup);
          rpcManager.sendToMany(owners, cmd, DeliverOrder.NONE);
-         stateToCleanup.forEach(this::removeStateFromLocal);
+         keysToCleanup.forEach(this::removeStateFromLocal);
       }
    }
 
```

Comment:
`removeStateFromLocal` is broken and never removes the state. A possible fix is below to remove if the `IracManagerKeyInfo` is equal.
```java
   private void removeStateFromLocal(IracManagerKeyInfo state) {
      updatedKeys.computeIfPresent(state.getKey(), (ignored, existingState) -> {
         // remove if the same state
         if (existingState.getKeyInfo().equals(state)) {
            if (log.isTraceEnabled()) {
               log.tracef("[IRAC] State removed? true, state=%s", existingState);
            }
            return null;
         } else {
            if (log.isTraceEnabled()) {
               log.tracef("[IRAC] State removed? false, state=%s", existingState);
            }
            return existingState;
         }
      });
   }
```